File: ios/Servantana/Sources/Models/AIModels.swift
import Foundation

// MARK: - AI Chat
struct AIChatRequest: Codable {
    let message: String
    let conversationHistory: [ChatMessage]

    init(message: String, history: [ChatMessage]) {
        self.message = message
        self.conversationHistory = history
    }
}

struct ChatMessage: Codable {
    let role: String
    let content: String
}

struct AIChatResponse: Codable {
    let message: String
    let suggestedActions: [SuggestedAction]?

    // Computed property for backward compatibility
    var response: String { message }
    var suggestions: [String]? {
        suggestedActions?.map { $0.label }
    }
}

struct SuggestedAction: Codable {
    let label: String
    let action: String
    let url: String?
}

// MARK: - Smart Match
struct SmartMatchRequest: Codable {
    let categoryId: String?
    let professionId: String?
    let latitude: Double
    let longitude: Double
    let maxDistance: Int
    let preferences: MatchPreferences?
}

struct MatchPreferences: Codable {
    let prioritizeRating: Bool?
    let prioritizePrice: Bool?
    let prioritizeDistance: Bool?
    let ecoFriendly: Bool?
    let petFriendly: Bool?
}

struct SmartMatchResponse: Codable {
    let matches: [SmartMatchResult]
    let totalCandidates: Int?
    let scoringWeights: [String: Double]?

    // Alias for backward compatibility
    var results: [SmartMatchResult] { matches }
}

struct SmartMatchResult: Codable, Identifiable {
    let worker: Worker
    let matchScore: Int
    let matchPercentage: Int
    let factors: MatchFactors
    let matchReasons: [String]

    var id: String { worker.id }

    // UI helper
    var recommendation: String? { matchReasons.first }
}

struct MatchFactors: Codable {
    let ratingScore: Double
    let experienceScore: Double
    let distanceScore: Double
    let priceScore: Double
    let availabilityScore: Double
    let preferencesScore: Double
    let reliabilityScore: Double
    let verificationScore: Double
    let responseTimeScore: Double
    let repeatCustomerScore: Double
}

// MARK: - Smart Schedule
struct SmartScheduleRequest: Codable {
    let date: String
    let professionId: String?
    let workerId: String?
    let duration: Int?
}

struct SmartScheduleResponse: Codable {
    let slots: [TimeSlot]
    let demandForecast: DemandForecast?
}

struct TimeSlot: Codable, Identifiable {
    let id: String
    let time: String
    let available: Bool
    let demandLevel: String
    let priceMultiplier: Double
    let recommendation: String?

    enum CodingKeys: String, CodingKey {
        case id, time, available, demandLevel, priceMultiplier, recommendation
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        time = try container.decode(String.self, forKey: .time)
        id = time
        available = try container.decodeIfPresent(Bool.self, forKey: .available) ?? true
        demandLevel = try container.decodeIfPresent(String.self, forKey: .demandLevel) ?? "normal"
        priceMultiplier = try container.decodeIfPresent(Double.self, forKey: .priceMultiplier) ?? 1.0
        recommendation = try container.decodeIfPresent(String.self, forKey: .recommendation)
    }
}

struct DemandForecast: Codable {
    let peakHours: [String]
    let lowDemandHours: [String]
    let recommendation: String
}

// MARK: - Review Insights
struct ReviewInsightsRequest: Codable {
    let workerId: String
}

struct ReviewInsightsResponse: Codable {
    let overallSentiment: String
    let trustScore: Double
    let insights: [ReviewInsight]
    let strengths: [String]
    let areasForImprovement: [String]
}

struct ReviewInsight: Codable, Identifiable {
    let category: String
    let sentiment: String
    let score: Double
    let mentions: Int
    let keywords: [String]

    var id: String { category }
}

// MARK: - Photo Analysis
struct PhotoAnalysisRequest: Codable {
    let imageUrls: [String]
    let analysisType: String // "single" or "before_after"
    let bookingId: String?

    init(imageUrls: [String], analysisType: String = "single", bookingId: String? = nil) {
        self.imageUrls = imageUrls
        self.analysisType = analysisType
        self.bookingId = bookingId
    }
}

struct PhotoAnalysisResponse: Codable {
    let type: String
    let summary: PhotoAnalysisSummary?
    let comparison: BeforeAfterComparison?
    let bookingId: String?
}

struct PhotoAnalysisSummary: Codable {
    let averageCleanlinessScore: Double
    let overallCondition: String
    let allConcerns: [String]
    let allPositives: [String]
    let averageJobComplexity: String
    let totalEstimatedTime: Int
    let averageConfidence: Int
}

struct BeforeAfterComparison: Codable {
    let improvementScore: Int
    let beforeScore: Int
    let afterScore: Int
    let improvements: [String]
    let remainingIssues: [String]
    let qualityVerified: Bool
    let verificationNotes: String
}

// MARK: - Price Estimate
struct PriceEstimateRequest: Codable {
    let images: [String]
    let propertyType: String
    let squareMeters: Double?
    let professionId: String?
}

struct PriceEstimateResponse: Codable {
    let estimatedPrice: PriceRange
    let estimatedDuration: DurationRange
    let factors: [PriceFactor]
    let confidence: Double
}

struct PriceRange: Codable {
    let min: Double
    let max: Double
    let currency: String
}

struct DurationRange: Codable {
    let min: Int
    let max: Int
    let unit: String
}

struct PriceFactor: Codable, Identifiable {
    let factor: String
    let impact: String
    let description: String

    var id: String { factor }
}

// MARK: - Route Optimization
struct RouteOptimizeRequest: Codable {
    let bookings: [RouteBooking]
    let startLocation: Location?
    let endLocation: Location?
}

struct RouteBooking: Codable {
    let bookingId: String
    let address: String
    let latitude: Double
    let longitude: Double
    let scheduledTime: String
    let duration: Int
}

struct Location: Codable {
    let latitude: Double
    let longitude: Double
    let address: String?
}

struct RouteOptimizeResponse: Codable {
    let optimizedRoute: [RouteStop]
    let totalDistance: Double
    let totalTime: Int
    let savedTime: Int
    let savedDistance: Double
}

struct RouteStop: Codable, Identifiable {
    let order: Int
    let bookingId: String
    let address: String
    let arrivalTime: String
    let departureTime: String
    let travelTimeToNext: Int?
    let distanceToNext: Double?

    var id: Int { order }
}

File: android/app/src/main/java/com/servantana/app/data/model/AIModels.kt
package com.servantana.app.data.model

import kotlinx.serialization.Serializable

// ==================== AI Chat ====================
@Serializable
data class AIChatRequest(
    val message: String,
    val conversationHistory: List<ChatMessage> = emptyList()
)

@Serializable
data class ChatMessage(
    val role: String, // "user" or "assistant"
    val content: String
)

@Serializable
data class AIChatResponse(
    val message: String,
    val suggestedActions: List<SuggestedAction> = emptyList(),
    val usage: TokenUsage? = null
)

@Serializable
data class SuggestedAction(
    val label: String,
    val action: String,
    val url: String? = null
)

@Serializable
data class TokenUsage(
    val inputTokens: Int,
    val outputTokens: Int
)

// UI-friendly chat message with additional metadata
data class AIChatMessage(
    val id: String,
    val role: String, // "user" or "assistant"
    val content: String,
    val timestamp: Long = System.currentTimeMillis()
)

// ==================== Smart Match ====================
@Serializable
data class SmartMatchRequest(
    val latitude: Double,
    val longitude: Double,
    val professionId: String? = null,
    val categoryId: String? = null,
    val preferredDate: String? = null,
    val preferredTime: String? = null,
    val maxPrice: Float? = null,
    val ecoFriendly: Boolean? = null,
    val petFriendly: Boolean? = null,
    val limit: Int = 10
)

@Serializable
data class SmartMatchResponse(
    val matches: List<WorkerMatch>,
    val totalCandidates: Int,
    val scoringWeights: Map<String, Float>? = null
)

@Serializable
data class WorkerMatch(
    val worker: Worker,
    val matchScore: Int,
    val matchPercentage: Int,
    val factors: MatchFactors,
    val matchReasons: List<String>
)

@Serializable
data class MatchFactors(
    val rating: Float = 0f,
    val experience: Float = 0f,
    val distance: Float = 0f,
    val price: Float = 0f,
    val availability: Float = 0f,
    val preferences: Float = 0f,
    val reliability: Float = 0f,
    val verification: Float = 0f,
    val responseTime: Float = 0f,
    val repeatCustomer: Float = 0f
)

// UI-friendly wrapper for smart match results
typealias SmartMatchResult = WorkerMatch

// Extension properties for UI compatibility
val SmartMatchResult.score: Int get() = matchPercentage
val SmartMatchResult.breakdown: Map<String, Double> get() = mapOf(
    "rating" to (factors.rating * 100).toDouble(),
    "experience" to (factors.experience * 100).toDouble(),
    "distance" to (factors.distance * 100).toDouble(),
    "price" to (factors.price * 100).toDouble(),
    "availability" to (factors.availability * 100).toDouble(),
    "preferences" to (factors.preferences * 100).toDouble(),
    "reliability" to (factors.reliability * 100).toDouble(),
    "verification" to (factors.verification * 100).toDouble(),
    "responseTime" to (factors.responseTime * 100).toDouble(),
    "repeatCustomer" to (factors.repeatCustomer * 100).toDouble()
)

// ==================== Smart Schedule ====================
@Serializable
data class SmartScheduleRequest(
    val workerId: String? = null,
    val professionId: String? = null,
    val categoryId: String? = null,
    val latitude: Double? = null,
    val longitude: Double? = null,
    val duration: Int = 120,
    val startDate: String? = null,
    val daysAhead: Int = 14
)

@Serializable
data class SmartScheduleResponse(
    val suggestions: List<TimeSlotSuggestion>,
    val categories: ScheduleCategories,
    val demandForecast: List<DemandPoint> = emptyList(),
    val totalSlotsAnalyzed: Int
)

@Serializable
data class TimeSlotSuggestion(
    val date: String,
    val time: String,
    val dayName: String,
    val score: Int,
    val priceModifier: Float,
    val demandLevel: String, // "low", "medium", "high", "peak"
    val reasons: List<String>,
    val estimatedWaitTime: Int,
    val availableWorkers: Int
)

@Serializable
data class ScheduleCategories(
    val bestValue: List<TimeSlotSuggestion> = emptyList(),
    val quickestConfirmation: List<TimeSlotSuggestion> = emptyList(),
    val mostAvailable: List<TimeSlotSuggestion> = emptyList()
)

@Serializable
data class DemandPoint(
    val dayOfWeek: Int,
    val hour: Int,
    val demand: Int
)

// ==================== Review Insights ====================
@Serializable
data class ReviewInsightsRequest(
    val workerId: String,
    val reviewIds: List<String>? = null,
    val analyzeAll: Boolean = true
)

@Serializable
data class ReviewInsightsResponse(
    val insights: ReviewInsights?,
    val stats: ReviewStats,
    val recentReviews: List<RecentReview> = emptyList()
)

@Serializable
data class ReviewInsights(
    val sentiment: SentimentAnalysis,
    val themes: List<ReviewTheme> = emptyList(),
    val strengths: List<String> = emptyList(),
    val weaknesses: List<String> = emptyList(),
    val trustScore: Int,
    val trustFactors: TrustFactors,
    val summary: String,
    val recommendations: List<String> = emptyList()
)

@Serializable
data class SentimentAnalysis(
    val overall: String, // "positive", "neutral", "negative"
    val score: Int, // -100 to 100
    val confidence: Int // 0-100
)

@Serializable
data class ReviewTheme(
    val theme: String,
    val sentiment: String,
    val mentions: Int,
    val examples: List<String> = emptyList()
)

@Serializable
data class TrustFactors(
    val authenticityScore: Int,
    val consistencyScore: Int,
    val detailScore: Int,
    val recencyScore: Int
)

@Serializable
data class ReviewStats(
    val totalReviews: Int,
    val averageRating: Float,
    val ratingsDistribution: List<Int>,
    val reviewsWithComments: Int,
    val avgReviewerAccountAge: Int,
    val avgReviewerBookings: Float
)

@Serializable
data class RecentReview(
    val id: String,
    val rating: Int,
    val comment: String?,
    val date: String,
    val reviewerName: String
)

// ==================== Photo Analysis ====================
@Serializable
data class PhotoAnalysisRequest(
    val imageUrls: List<String>,
    val analysisType: String = "single", // "single" or "before_after"
    val bookingId: String? = null
)

@Serializable
data class PhotoAnalysisResponse(
    val type: String,
    val photos: List<PhotoAnalysis>? = null,
    val comparison: BeforeAfterComparison? = null,
    val summary: PhotoSummary? = null,
    val bookingId: String? = null
)

@Serializable
data class PhotoAnalysis(
    val cleanlinessScore: Int,
    val overallCondition: String,
    val areasOfConcern: List<String>,
    val positiveAspects: List<String>,
    val jobComplexity: String,
    val estimatedTime: Int,
    val recommendations: List<String>,
    val confidence: Int
)

@Serializable
data class BeforeAfterComparison(
    val improvementScore: Int,
    val beforeScore: Int,
    val afterScore: Int,
    val improvements: List<String>,
    val remainingIssues: List<String>,
    val qualityVerified: Boolean,
    val verificationNotes: String
)

@Serializable
data class PhotoSummary(
    val averageCleanlinessScore: Float,
    val overallCondition: String,
    val allConcerns: List<String>,
    val allPositives: List<String>,
    val averageJobComplexity: String,
    val totalEstimatedTime: Int,
    val averageConfidence: Int
)

// ==================== Price Estimate ====================
@Serializable
data class PriceEstimateRequest(
    val imageUrls: List<String>,
    val serviceType: String = "cleaning",
    val professionId: String? = null,
    val additionalInfo: String? = null,
    val userCurrency: String = "USD"
)

@Serializable
data class PriceEstimateResponse(
    val estimate: PriceEstimate,
    val marketData: MarketData,
    val comparableBookings: List<ComparableBooking> = emptyList()
)

@Serializable
data class PriceEstimate(
    val estimatedPrice: PriceRange,
    val breakdown: PriceBreakdown,
    val spaceAnalysis: SpaceAnalysis,
    val timeEstimate: TimeEstimate,
    val specialRequirements: List<String>,
    val confidence: Int,
    val notes: String
)

@Serializable
data class PriceRange(
    val low: Int,
    val mid: Int,
    val high: Int,
    val currency: String
)

@Serializable
data class PriceBreakdown(
    val basePrice: Int,
    val sizeMultiplier: Float,
    val difficultyMultiplier: Float,
    val specialtyAddons: Int
)

@Serializable
data class SpaceAnalysis(
    val estimatedSqMeters: Int,
    val roomCount: Int,
    val roomTypes: List<String>,
    val condition: String,
    val difficulty: Int
)

@Serializable
data class TimeEstimate(
    val minMinutes: Int,
    val maxMinutes: Int,
    val recommended: Int
)

@Serializable
data class MarketData(
    val avgHourlyRate: Int,
    val rateRange: RateRange,
    val currency: String
)

@Serializable
data class RateRange(
    val min: Int,
    val max: Int
)

@Serializable
data class ComparableBooking(
    val price: Float,
    val duration: Int,
    val service: String?
)

// ==================== Route Optimize ====================
@Serializable
data class RouteOptimizeRequest(
    val date: String,
    val bookingIds: List<String>? = null,
    val startLocation: Location? = null
)

@Serializable
data class Location(
    val latitude: Double,
    val longitude: Double
)

@Serializable
data class RouteOptimizeResponse(
    val route: OptimizedRoute,
    val metadata: RouteMetadata
)

@Serializable
data class OptimizedRoute(
    val originalOrder: List<RouteLocation>,
    val optimizedOrder: List<RouteLocation>,
    val savings: RouteSavings,
    val totalDistance: Float,
    val totalDuration: Int,
    val legs: List<RouteLeg>,
    val schedule: List<ScheduleEntry>
)

@Serializable
data class RouteLocation(
    val id: String,
    val name: String,
    val latitude: Double,
    val longitude: Double,
    val scheduledTime: String,
    val duration: Int,
    val address: String? = null
)

@Serializable
data class RouteSavings(
    val distanceKm: Float,
    val estimatedMinutes: Int,
    val percentImprovement: Int
)

@Serializable
data class RouteLeg(
    val from: String,
    val to: String,
    val distanceKm: Float,
    val estimatedMinutes: Int
)

@Serializable
data class ScheduleEntry(
    val bookingId: String,
    val arrivalTime: String,
    val departureTime: String,
    val address: String
)

@Serializable
data class RouteMetadata(
    val bookingsOptimized: Int,
    val date: String,
    val startLocation: Location?
)
